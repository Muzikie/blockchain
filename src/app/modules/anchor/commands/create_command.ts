/* eslint-disable @typescript-eslint/member-ordering */
/* eslint-disable class-methods-use-this */
import {
  BaseCommand,
  CommandVerifyContext,
  CommandExecuteContext,
  VerificationResult,
  VerifyStatus,
} from 'lisk-sdk';
import { AnchorStore } from '../stores/anchor';
import { AnchorAccountStore } from '../stores/anchorAccount';
import { AnchorCreated } from '../events/anchorCreated';
import { CreateCommandParams, Anchor, AnchorAccount } from '../types';
import { createCommandParamsSchema } from '../schemas';
import { getCreatedAt, getAnchorID } from '../utils';
import { VOTE_RATE_LIMIT } from '../constants';
import { BadgeMethod } from '../../badge/method';
import { Badges } from '../../badge/types';
import { getBadgeID } from '../../badge/utils';
import { AnchorStatsStore } from '../stores/anchorStats';

export class CreateCommand extends BaseCommand {
  public schema = createCommandParamsSchema;
  private _badgeMethod!: BadgeMethod;

  public addDependencies(badgeMethod: BadgeMethod) {
    this._badgeMethod = badgeMethod;
  }

  // eslint-disable-next-line @typescript-eslint/require-await
  public async verify(
    context: CommandVerifyContext<CreateCommandParams>,
  ): Promise<VerificationResult> {
    const anchorStore = this.stores.get(AnchorStore);
    const { senderAddress } = context.transaction;
    const anchorAccountStore = this.stores.get(AnchorAccountStore);
    const anchorId = getAnchorID(context.params);

    const anchorExists = await anchorStore.has(context, anchorId);
    if (anchorExists) {
      throw new Error('This anchor already exist.');
    }

    // Add submission rate limit
    const senderExists = await anchorAccountStore.has(context, senderAddress);
    if(senderExists) {
      const senderAccount = await anchorAccountStore.get(context, senderAddress);
      const IDS = senderAccount.anchors.slice(-1 * VOTE_RATE_LIMIT);

      if(IDS.length >= VOTE_RATE_LIMIT) {
        const anchor = await anchorStore.get(context, IDS[0]);

        if (anchor.createdAt === getCreatedAt(Math.floor(((new Date()).getTime())))) {
          throw new Error(`You have exceeded the ${VOTE_RATE_LIMIT} anchor submissions daily limit.`);
        }
      }
    }

    return { status: VerifyStatus.OK };
  }

  public async execute(context: CommandExecuteContext<CreateCommandParams>): Promise<void> {
    const {
      params,
      transaction: { senderAddress },
    } = context;
    const anchorAccountStore = this.stores.get(AnchorAccountStore);
    const anchorStore = this.stores.get(AnchorStore);
    const anchorStatsStore = this.stores.get(AnchorStatsStore);
    const methodContext = context.getMethodContext();

    // Create anchor ID
    const anchorID = getAnchorID(context.params);
    const createdAt = getCreatedAt(new Date().getTime());
    // Create anchor object
    const anchor: Anchor = {
      ...params,
      id: anchorID,
      createdAt,
      votes: [],
      submitter: senderAddress,
    };

    const badgeIDs = [1, 2, 3]
      .map((rank) => getBadgeID(createdAt, rank, Badges.AOTD));
    context.logger.info(badgeIDs.map((idd) => idd.toString('hex')));
    // Store the anchor object in the blockchain
    await anchorStore.set(context, anchorID, anchor);

    const StatsExist = await anchorStatsStore.has(context, Buffer.from(createdAt));
  
    if (!StatsExist) {
      const newAnchorStats = {
        date: createdAt,
        anchorsCount: 1,
        votesCount: 0,
      };
      await anchorStatsStore.set(context, Buffer.from(createdAt), newAnchorStats);
      context.logger.info(`Anchors count for ${createdAt}: 1`);
    } else {
      const anchorStats = await anchorStatsStore.get(context, Buffer.from(createdAt));
      anchorStats.anchorsCount += 1;
      await anchorStatsStore.set(context, Buffer.from(createdAt), anchorStats);
      context.logger.info(`Anchors count2SS for ${createdAt}: ${anchorStats.anchorsCount}`);
    }

    // Get the sender account from the blockchain
    const senderExists = await anchorAccountStore.has(context, senderAddress);
    let senderAccount: AnchorAccount;
    if (!senderExists) {
      senderAccount = {
        anchors: [anchorID],
        votes: [],
      };
    } else {
      const retrievedAccount = await anchorAccountStore.get(context, senderAddress);
      senderAccount = {
        anchors: [...retrievedAccount.anchors, anchorID],
        votes: retrievedAccount.votes
      };
    }

    // Store the account object in the blockchain
    await anchorAccountStore.set(context, senderAddress, senderAccount);

    await this._badgeMethod.createBadgesForDay(methodContext, createdAt);

    const anchorCreated = this.events.get(AnchorCreated);
    anchorCreated.add(context, {
      submitter: context.transaction.senderAddress,
      anchorID,
      createdAt,
    }, [context.transaction.senderAddress]);
  }
}
