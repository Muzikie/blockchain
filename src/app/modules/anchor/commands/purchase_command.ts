/* eslint-disable @typescript-eslint/member-ordering */
/* eslint-disable class-methods-use-this */

import {
  BaseCommand,
  CommandVerifyContext,
  CommandExecuteContext,
  VerificationResult,
  VerifyStatus,
  TokenMethod,
} from 'lisk-sdk';
import { AnchorStore } from '../stores/anchor';
import { AnchorAccountStore } from '../stores/anchorAccount';
import { PurchaseCommandParams, AnchorAccount } from '../types';
import { purchaseCommandParamsSchema } from '../schemas';
import { DEV_ADDRESS, TREASURY_ADDRESS } from '../constants';

export class PurchaseCommand extends BaseCommand {
  public schema = purchaseCommandParamsSchema;
  private _tokenMethod!: TokenMethod;

  public addDependencies(tokenMethod: TokenMethod) {
    this._tokenMethod = tokenMethod;
  }

  // eslint-disable-next-line @typescript-eslint/require-await
  public async verify(
    context: CommandVerifyContext<PurchaseCommandParams>,
  ): Promise<VerificationResult> {
    if (context.params.members?.length === 0) {
      return {
        status: VerifyStatus.FAIL,
        error: new Error('Define at least one member'),
      };
    }
    return { status: VerifyStatus.OK };
  }

  public async execute(context: CommandExecuteContext<PurchaseCommandParams>): Promise<void> {
    const {
      params: { members, anchorID },
      transaction: { senderAddress },
      chainID,
    } = context;
    const tokenID = Buffer.concat([chainID, Buffer.alloc(4)]);
    const methodContext = context.getMethodContext();
    const anchorAccountStore = this.stores.get(AnchorAccountStore);
    const anchorStore = this.stores.get(AnchorStore);

    // Throw if anchor didn't exist
    const anchorExists = await anchorStore.has(context, anchorID);
    if (!anchorExists) {
      throw new Error(`Anchor with ID ${anchorID.toString('hex')} does not exist`);
    }
    // Get anchor from the blockchain
    const anchorNFT = await anchorStore.get(context, anchorID);
    // Throw error if the members are more than the maximum allowed
    if (members.length > anchorNFT.maxMembers) {
      throw new Error(`The anchor only allows ${anchorNFT.maxMembers} members`);
    }
    // Throw error if the anchor doesn't is already active
    if (!anchorNFT.creatorAddress.equals(DEV_ADDRESS)) {
      throw new Error(`The anchor is already purchased.`);
    }

    // throw an error if the sender does not have enough tokens
    // Deduct the price from the sender account
    // Add 20% of the price to the creator account
    const devCosts = (anchorNFT.price * BigInt(2)) / BigInt(10);
    const consumable = anchorNFT.price - devCosts;

    await this._tokenMethod.transfer(
      methodContext,
      senderAddress,
      DEV_ADDRESS,
      tokenID,
      devCosts,
    );
    await this._tokenMethod.transfer(
      methodContext,
      senderAddress,
      TREASURY_ADDRESS,
      tokenID,
      consumable,
    );

    // Create anchor object
    const anchor = {
      ...anchorNFT,
      consumable,
      members,
      creatorAddress: senderAddress,
    };
    // Save anchor object on the blockchain
    await anchorStore.set(context, anchorID, anchor);

    // Add owned anchor and save the anchor on the sender account
    const senderExist = await anchorAccountStore.has(context, senderAddress);
    let senderAccount: AnchorAccount;
    if (senderExist) {
      senderAccount = await anchorAccountStore.get(context, senderAddress);
      senderAccount.anchor.owned = [...senderAccount.anchor.owned, anchorID];
    } else {
      senderAccount = {
        anchor: {
          owned: [anchorID],
          shared: Buffer.alloc(0),
        },
      };
    }

    await anchorAccountStore.set(context, senderAddress, senderAccount);

    // Remove anchor from the dev account
    const devAccount = await anchorAccountStore.get(context, DEV_ADDRESS);
    await anchorAccountStore.set(
      context,
      DEV_ADDRESS,
      {
        anchor: {
          owned: devAccount.anchor.owned.filter(item => !item.equals(anchorID)),
          shared: Buffer.alloc(0),
        },
      },
    );

    // Save the anchor on the members accounts
    for (const member of members) {
      const memberExist = await anchorAccountStore.has(context, member);
      let memberAccount: AnchorAccount;
      if (memberExist) {
        memberAccount = await anchorAccountStore.get(context, member);
        memberAccount.anchor.shared = anchorID;
      } else {
        memberAccount = {
          anchor: {
            owned: [Buffer.alloc(0)],
            shared: anchorID,
          },
        };
      }
      await anchorAccountStore.set(context, member, memberAccount);
    }
  }
}
