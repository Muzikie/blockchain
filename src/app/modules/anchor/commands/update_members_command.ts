/* eslint-disable @typescript-eslint/member-ordering */
/* eslint-disable class-methods-use-this */

import {
  BaseCommand,
  CommandVerifyContext,
  CommandExecuteContext,
  VerificationResult,
  VerifyStatus,
} from 'lisk-sdk';
import { AnchorStore } from '../stores/anchor';
import { AnchorAccountStore } from '../stores/anchorAccount';
import { UpdateMembersCommandParams, AnchorAccount } from '../types';
import { updateMembersCommandParamsSchema } from '../schemas';

export class UpdateMembersCommand extends BaseCommand {
  public schema = updateMembersCommandParamsSchema;

  // eslint-disable-next-line @typescript-eslint/require-await
  public async verify(
    context: CommandVerifyContext<UpdateMembersCommandParams>,
  ): Promise<VerificationResult> {
    if (context.params.members?.length === 0) {
      return {
        status: VerifyStatus.FAIL,
        error: new Error('Define at least one member'),
      };
    }
    return { status: VerifyStatus.OK };
  }

  public async execute(context: CommandExecuteContext<UpdateMembersCommandParams>): Promise<void> {
    const { members, anchorID } = context.params;
    const { senderAddress } = context.transaction;
    const anchorAccountStore = this.stores.get(AnchorAccountStore);
    const anchorStore = this.stores.get(AnchorStore);

    // Throw if anchor didn't exist
    const anchorExists = await anchorStore.has(context, anchorID);
    if (!anchorExists) {
      throw new Error(`Anchor with ID ${anchorID.toString('hex')} does not exist`);
    }
    // Get anchor from the blockchain
    const anchorNFT = await anchorStore.get(context, anchorID);
    // Throw error if the members are more than the maximum allowed
    if (members.length > anchorNFT.maxMembers) {
      throw new Error(`The anchor only allows ${anchorNFT.maxMembers} members`);
    }
    // Throw error if the anchor doesn't is already active
    if (!anchorNFT.creatorAddress.equals(senderAddress)) {
      throw new Error('Only the owner of the anchor can update the members.');
    }

    const oldMembers = anchorNFT.members;

    // Create anchor object
    const anchor = {
      ...anchorNFT,
      members,
    };

    // Save anchor object on the blockchain
    await anchorStore.set(context, anchorID, anchor);

    let addedMembers = [...members];
    // Remove the anchor from the members accounts who are now removed
    for (const oldMember of oldMembers) {
      const unChanged = members.find(item => item.equals(oldMember));
      if (!unChanged) {
        // Remove the anchor ID from the account's shared anchors
        const oldMemberAccount = await anchorAccountStore.get(context, oldMember);
        oldMemberAccount.anchor.shared = Buffer.alloc(0);
        await anchorAccountStore.set(context, oldMember, oldMemberAccount);
      } else {
        addedMembers = addedMembers.filter(item => !item.equals(oldMember));
      }
    }

    // Add anchor ID to the new member accounts
    for (const member of addedMembers) {
      const memberExist = await anchorAccountStore.has(context, member);
      let memberAccount: AnchorAccount;
      if (memberExist) {
        memberAccount = await anchorAccountStore.get(context, member);
        memberAccount.anchor.shared = anchorID;
      } else {
        memberAccount = {
          anchor: {
            owned: [Buffer.alloc(0)],
            shared: anchorID,
          },
        };
      }
      await anchorAccountStore.set(context, member, memberAccount);
    }
  }
}
