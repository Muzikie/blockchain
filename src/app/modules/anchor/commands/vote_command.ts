/* eslint-disable @typescript-eslint/member-ordering */
/* eslint-disable class-methods-use-this */

import {
  BaseCommand,
  CommandVerifyContext,
  CommandExecuteContext,
  VerificationResult,
  VerifyStatus,
  TokenMethod,
} from 'lisk-sdk';
import { AnchorStore } from '../stores/anchor';
import { AnchorAccountStore } from '../stores/anchorAccount';
import { VoteCommandParams, AnchorAccount } from '../types';
import { voteCommandParamsSchema } from '../schemas';
import { DEV_ADDRESS, TREASURY_ADDRESS, CONTRIBUTION_FEE } from '../constants';

export class VoteCommand extends BaseCommand {
  public schema = voteCommandParamsSchema;
  private _tokenMethod!: TokenMethod;

  public addDependencies(tokenMethod: TokenMethod) {
    this._tokenMethod = tokenMethod;
  }

  // eslint-disable-next-line @typescript-eslint/require-await
  public async verify(
    context: CommandVerifyContext<VoteCommandParams>,
  ): Promise<VerificationResult> {
    const {
      params: { anchorID },
    } = context;
    const anchorStore = this.stores.get(AnchorStore);
    // Throw if anchor didn't exist
    const anchorExists = await anchorStore.has(context, anchorID);
    if (!anchorExists) {
      throw new Error(`Anchor with ID ${anchorID.toString('hex')} does not exist`);
    }

    // @todo Throw error if already voted

    // @todo Add vote rate limit

    return { status: VerifyStatus.OK };
  }

  public async execute(context: CommandExecuteContext<VoteCommandParams>): Promise<void> {
    const {
      params: { anchorID },
      transaction: { senderAddress },
      chainID,
    } = context;
    const tokenID = Buffer.concat([chainID, Buffer.alloc(4)]);
    const methodContext = context.getMethodContext();
    const anchorAccountStore = this.stores.get(AnchorAccountStore);
    const anchorStore = this.stores.get(AnchorStore);

    // Get anchor from the blockchain and add the sender address to the votes
    const anchorNFT = await anchorStore.get(context, anchorID);

    // Collect the contribution fee
    await this._tokenMethod.transfer(
      methodContext,
      senderAddress,
      DEV_ADDRESS,
      tokenID,
      CONTRIBUTION_FEE,
    );

    // Create anchor object
    const updatedAnchor = {
      ...anchorNFT,
      votes: [...anchorNFT.votes, senderAddress],
    };
    // Save anchor object on the blockchain
    await anchorStore.set(context, anchorID, updatedAnchor);

    // Add owned anchor and save the anchor on the sender account
    const senderExist = await anchorAccountStore.has(context, senderAddress);
    let senderAccount: AnchorAccount;
    if (senderExist) {
      senderAccount = await anchorAccountStore.get(context, senderAddress);
      senderAccount.votes = [...senderAccount.votes, anchorID];
    } else {
      senderAccount = {
        anchors: [],
        votes: [anchorID],
      };
    }

    await anchorAccountStore.set(context, senderAddress, senderAccount);
  }
}
