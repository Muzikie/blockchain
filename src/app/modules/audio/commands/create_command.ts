/* eslint-disable class-methods-use-this */
import {
  BaseCommand,
  CommandVerifyContext,
  CommandExecuteContext,
  VerificationResult,
  VerifyStatus,
} from 'lisk-sdk';
import { AudioStore } from '../stores/audio';
import { AudioAccountStore } from '../stores/audioAccount';
import { CreateCommandParams, Audio, AudioAccount } from '../types';
import { createCommandParamsSchema } from '../schemas';
import { VALID_GENRES, MIN_RELEASE_YEAR } from '../constants';
import { getEntityID, verifyHash } from '../../../utils';
import { AudioCreated } from '../events/audioCreated'

export class CreateCommand extends BaseCommand {
  public schema = createCommandParamsSchema;

  // eslint-disable-next-line @typescript-eslint/require-await
  public async verify(
    context: CommandVerifyContext<CreateCommandParams>,
  ): Promise<VerificationResult> {
    const { params, transaction } = context;
    const thisYear = new Date().getFullYear();
    const numericYear = Number(params.releaseYear);
    if (numericYear < MIN_RELEASE_YEAR || numericYear > thisYear) {
      return {
        status: VerifyStatus.FAIL,
        error: new Error(
          `Release year must be a number between ${MIN_RELEASE_YEAR} and ${thisYear}`,
        ),
      };
    }
    if (params.genre.some(item => item > VALID_GENRES.length)) {
      return {
        status: VerifyStatus.FAIL,
        error: new Error('Genres should be selected from the list of valid genres'),
      };
    }
    if (
      params.owners.length === 0 ||
      params.owners.some(item => item.shares < 1) ||
      params.owners.reduce((acc, item) => acc + item.shares, 0) !== 100
    ) {
      return {
        status: VerifyStatus.FAIL,
        error: new Error(
          'Owners should have a total of 100 shares. Each owner should have a positive number of shares.',
        ),
      };
    }

    const isHashGenuine = verifyHash(params.audioSignature, params.audioHash, transaction.senderPublicKey);
    if (!isHashGenuine) {
      return {
        status: VerifyStatus.FAIL,
        error: new Error('The hash signature is not generated by the sender'),
      };
    }

    return { status: VerifyStatus.OK };
  }

  public async execute(context: CommandExecuteContext<CreateCommandParams>): Promise<void> {
    const { params, transaction } = context;
    // Get namehash output of the audio file
    const audioID = getEntityID(context.transaction);

    const audioAccountSubStore = this.stores.get(AudioAccountStore);
    const audioSubStore = this.stores.get(AudioStore);

    // Check uniqueness of the NFT
    const audioExists = await audioSubStore.has(context, audioID);
    if (audioExists) {
      throw new Error('You have already created this audio.');
    }

    // Create the Audio object and save it on the blockchain
    const owners = params.owners.map(owner => ({
      address: owner.address,
      shares: owner.shares,
      income: BigInt(0),
    }));

    const audioObject: Audio = {
      ...params,
      owners,
      creatorAddress: transaction.senderAddress,
    };

    // Store the hash of the audio object in the sender account
    const accountExists = await audioAccountSubStore.has(context, transaction.senderAddress);
    if (accountExists) {
      const senderAccount: AudioAccount = await audioAccountSubStore.get(
        context,
        transaction.senderAddress,
      );
      senderAccount.audio.audios = [...senderAccount.audio.audios, audioID];
      await audioAccountSubStore.set(context, transaction.senderAddress, senderAccount);
    } else {
      await audioAccountSubStore.set(context, context.transaction.senderAddress, {
        audio: {
          audios: [audioID],
        },
      });
    }

    // @todo Here we should check if the Audio is already uploaded using steganography methods
    // Store the audio object in the blockchain
    await audioSubStore.set(context, audioID, audioObject);

    // Emit a "New collection" event
    const audioCreated = this.events.get(AudioCreated);
    audioCreated.add(context, {
      creatorAddress: context.transaction.senderAddress,
      audioID,
    }, [context.transaction.senderAddress]);
  }
}
