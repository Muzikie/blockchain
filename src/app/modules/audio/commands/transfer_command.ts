/* eslint-disable class-methods-use-this */
import {
  BaseCommand,
  CommandVerifyContext,
  CommandExecuteContext,
  VerificationResult,
  VerifyStatus,
} from 'lisk-sdk';
import { AudioStore } from '../stores/audio';
import { AudioAccountStore } from '../stores/audioAccount';
import { TransferCommandParams, AudioAccount } from '../types';
import { transferCommandParamsSchema } from '../schemas';

export class TransferCommand extends BaseCommand {
  public schema = transferCommandParamsSchema;

  // eslint-disable-next-line @typescript-eslint/require-await
  public async verify(
    _context: CommandVerifyContext<TransferCommandParams>,
  ): Promise<VerificationResult> {
    return { status: VerifyStatus.OK };
  }

  public async execute(context: CommandExecuteContext<TransferCommandParams>): Promise<void> {
    const audioAccountSubStore = this.stores.get(AudioAccountStore);
    const audioSubStore = this.stores.get(AudioStore);

    // Get audio NFT and both accounts from substores
    const { address, audioID, shares } = context.params;
    const { senderAddress } = context.transaction;
    const audioExists = await audioSubStore.has(context, audioID);

    if (!audioExists) {
      throw new Error('Audio NFT does not exist.');
    }

    const audioNFT = await audioSubStore.get(context, audioID);
    const senderShare = audioNFT.owners.find(item => item.address.equals(senderAddress));

    if (!senderShare) {
      throw new Error('You do not own share of this audio.');
    }

    if (senderShare.shares < shares) {
      throw new Error(`You may only transfer 1-${senderShare.shares} shares (your shares)`);
    }

    const audioCreator = await audioAccountSubStore.get(context, audioNFT.creatorAddress);
    const recipientExists = await audioAccountSubStore.has(context, address);
    let oldIncome = BigInt(0);

    // set old owner shares = old shares - context.params.shares
    audioNFT.owners = audioNFT.owners
      .map(item => {
        if (item.address.equals(senderAddress)) {
          oldIncome = item.income;
          return {
            ...item,
            shares: item.shares - shares,
          };
        }
        return item;
      })
      // if old owner shares = 0, remove old owner from owners array
      .filter(item => item.shares > 0);

    // if old owner shares = 0, remove audio ID from old owner's ownedAudio array
    if (senderShare.shares === shares) {
      audioCreator.audio.audios = audioCreator.audio.audios.filter(item => !item.equals(audioID));
    }
    // set new owner shares = context.params.shares
    let recipientAccount: AudioAccount;
    if (recipientExists) {
      recipientAccount = await audioAccountSubStore.get(context, address);
      if (!recipientAccount.audio.audios.find(item => item.equals(audioID))) {
        recipientAccount.audio.audios.push(audioID);
      }
    } else {
      recipientAccount = {
        audio: {
          audios: [audioID],
        },
      };
    }
    await audioAccountSubStore.set(context, address, recipientAccount);

    const recipientIndex = audioNFT.owners.findIndex(item => item.address.equals(address));
    if (recipientIndex === -1) {
      // if recipient does not exist, add new owner to owners array
      audioNFT.owners.push({
        address,
        shares,
        income: senderShare.shares === shares ? BigInt(0) : oldIncome, // income is 0 for new owners
      });
    } else {
      // if recipient exists, add new owner shares to recipient shares
      audioNFT.owners[recipientIndex].shares += shares;
    }

    // update audio NFT in audioSubStore
    await audioSubStore.set(context, audioID, audioNFT);
    // update audio account in audioAccountSubStore
    await audioAccountSubStore.set(context, senderAddress, audioCreator);
  }
}
