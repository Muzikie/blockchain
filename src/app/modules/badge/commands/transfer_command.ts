/* eslint-disable @typescript-eslint/member-ordering */
/* eslint-disable class-methods-use-this */
import {
  BaseCommand,
  CommandVerifyContext,
  CommandExecuteContext,
  VerificationResult,
  VerifyStatus,
} from 'lisk-sdk';
import { BadgeStore } from '../stores/badge';
import { BadgeAccountStore } from '../stores/badgeAccount';
import { TransferCommandParams, BadgeAccount } from '../types';
import { transferCommandParamsSchema } from '../schemas';

export class TransferCommand extends BaseCommand {
  public schema = transferCommandParamsSchema;

  // eslint-disable-next-line @typescript-eslint/require-await
  public async verify(
    _context: CommandVerifyContext<TransferCommandParams>,
  ): Promise<VerificationResult> {
    return { status: VerifyStatus.OK };
  }

  public async execute(context: CommandExecuteContext<TransferCommandParams>): Promise<void> {
    const badgeAccountSubStore = this.stores.get(BadgeAccountStore);
    const badgeSubStore = this.stores.get(BadgeStore);

    // Get badge NFT and both accounts from substores
    const { address, badgeID, shares } = context.params;
    const { senderAddress } = context.transaction;
    const badgeExists = await badgeSubStore.has(context, badgeID);

    if (!badgeExists) {
      throw new Error('Badge NFT does not exist.');
    }

    const badgeNFT = await badgeSubStore.get(context, badgeID);
    const senderShare = badgeNFT.owners.find(item => item.address.equals(senderAddress));

    if (!senderShare) {
      throw new Error('You do not own share of this badge.');
    }

    if (senderShare.shares < shares) {
      throw new Error(`You may only transfer ${senderShare.shares} shares (your shares)`);
    }

    const recipientExists = await badgeAccountSubStore.has(context, address);
    let oldIncome = BigInt(0);

    // set old owner shares = old shares - context.params.shares
    badgeNFT.owners = badgeNFT.owners
      .map(item => {
        if (item.address.equals(senderAddress)) {
          oldIncome = item.income;
          return {
            ...item,
            shares: item.shares - shares,
          };
        }
        return item;
      })
      // if old owner shares = 0, remove old owner from owners array
      .filter(item => item.shares > 0);

    // if old owner shares = 0, remove badge ID from old owner's ownedBadge array
    const senderAccount = await badgeAccountSubStore.get(context, senderAddress);
    if (senderShare.shares === shares) {
      senderAccount.badge.badges = senderAccount.badge.badges.filter(item => !item.equals(badgeID));
      // update badge account in badgeAccountSubStore
      await badgeAccountSubStore.set(context, senderAddress, senderAccount);
    }

    // set new owner shares = context.params.shares
    let recipientAccount: BadgeAccount;
    if (recipientExists) {
      recipientAccount = await badgeAccountSubStore.get(context, address);
      if (!recipientAccount.badge.badges.find(item => item.equals(badgeID))) {
        recipientAccount.badge.badges.push(badgeID);
      }
    } else {
      recipientAccount = {
        badge: {
          badges: [badgeID],
        },
      };
    }
    await badgeAccountSubStore.set(context, address, recipientAccount);

    const recipientIndex = badgeNFT.owners.findIndex(item => item.address.equals(address));
    if (recipientIndex === -1) {
      // if recipient does not exist, add new owner to owners array
      badgeNFT.owners.push({
        address,
        shares,
        income: senderShare.shares === shares ? oldIncome : BigInt(0), // income is 0 for new owners
      });
    } else {
      // if recipient exists, add new owner shares to recipient shares
      badgeNFT.owners[recipientIndex].shares += shares;
    }

    // update badge NFT in badgeSubStore
    await badgeSubStore.set(context, badgeID, badgeNFT);
  }
}
