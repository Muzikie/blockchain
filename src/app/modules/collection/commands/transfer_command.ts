/* eslint-disable class-methods-use-this */

import {
  BaseCommand,
  CommandVerifyContext,
  CommandExecuteContext,
  VerificationResult,
  VerifyStatus,
} from 'lisk-sdk';
import { CollectionStore } from '../stores/collection';
import { CollectionAccountStore } from '../stores/collectionAccount';
import { TransferCommandParams, CollectionAccount } from '../types';
import { transferCommandParamsSchema } from '../schemas';

export class TransferCommand extends BaseCommand {
  public schema = transferCommandParamsSchema;

  // eslint-disable-next-line @typescript-eslint/require-await
  public async verify(_context: CommandVerifyContext<TransferCommandParams>): Promise<VerificationResult> {
    return { status: VerifyStatus.OK };
  }

  public async execute(context: CommandExecuteContext<TransferCommandParams>): Promise<void> {
    const collectionAccountSubStore = this.stores.get(CollectionAccountStore);
    const collectionSubStore = this.stores.get(CollectionStore);

    // Get collection NFT and both accounts from substores
    const { address, collectionID } = context.params;
    const collectionExists = await collectionSubStore.has(context, collectionID);

    if (!collectionExists) {
      throw new Error('Audio NFT does not exist.');
    }

    const collectionNFT = await collectionSubStore.get(context, collectionID);

    if (!collectionNFT.ownerAddress.equals(context.transaction.senderAddress)) {
      throw new Error('You are not the owner of this collection.');
    }

    const collectionAccount = await collectionAccountSubStore.get(context, collectionNFT.ownerAddress);
    const recipientExists = await collectionAccountSubStore.has(context, address);

    // Find and remove the collectionID from the owner account
    const collectionIndex = collectionAccount.collection.collections.findIndex(id => id.equals(collectionID));
    if (collectionIndex === -1) {
      throw new Error('Collection not found in the sender account.');
    }
    collectionAccount.collection.collections.splice(collectionIndex, 1);

    // Add the collectionID to the recipient account
    collectionNFT.ownerAddress = address;
    if (recipientExists) {
      const recipientAccount = await collectionAccountSubStore.get(context, address);
      recipientAccount.collection.collections.push(collectionID);
      await collectionAccountSubStore.set(context, address, recipientAccount);
    } else {
      const recipientAccount: CollectionAccount = {
        collection: {
          collections: [collectionID],
        }
      };
      await collectionAccountSubStore.set(context, address, recipientAccount);
    }

    // Update the sender account on the blockchain
    await collectionAccountSubStore.set(context, context.transaction.senderAddress, collectionAccount);
    // Update the NFT token on the blockchain
    await collectionSubStore.set(context, collectionID, collectionNFT);
  }
}
