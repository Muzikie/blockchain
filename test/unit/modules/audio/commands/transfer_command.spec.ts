import { TransferCommand } from '../../../../../src/app/modules/audio/commands/transfer_command';

describe('TransferCommand', () => {
  let command: TransferCommand;

  beforeEach(() => {
    command = new TransferCommand();
  });

  describe('constructor', () => {
    it('should have valid name', () => {
      expect(command.name).toEqual('transfer');
    });

    it('should have valid schema', () => {
      expect(command.schema).toMatchSnapshot();
    });
  });

  describe('verify', () => {
    describe('schema validation', () => {
      it.todo('should throw errors for invalid schema');
      it.todo('should be ok for valid schema');
    });
  });

  describe('execute', () => {
    describe('valid cases', () => {
      it.todo('should update the state store');
    });

    describe('invalid cases', () => {
      it.todo('should throw error');
    });
  });
});
