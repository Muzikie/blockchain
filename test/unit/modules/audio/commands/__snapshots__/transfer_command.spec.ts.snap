// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TransferCommand constructor should have valid schema 1`] = `
Object {
  "$id": "audio/transfer",
  "properties": Object {
    "address": Object {
      "dataType": "bytes",
      "fieldNumber": 2,
      "format": "lisk32",
    },
    "audioID": Object {
      "dataType": "bytes",
      "fieldNumber": 1,
    },
    "shares": Object {
      "dataType": "uint32",
      "fieldNumber": 3,
    },
  },
  "required": Array [
    "audioID",
    "address",
  ],
  "title": "TransferAsset transaction asset for audio module",
  "type": "object",
}
`;
